api/bots_api/bots_test_lib.py

```diff
@@ -56,42 +56,99 @@ def initialize_bot(self):
+            'sender_full_name': "Foo bar",
+        },
+        'stream': {
+            'type': 'stream',
+            'display_recipient': 'foo',  # stream name
+            'subject': 'foo',            # topic name
+            'sender_email': "foo_sender@zulip.com",
+            'sender_id': 0,
+            'sender_full_name': "Foo Bar",
+        },
+    }
+    response_template = {
+        'private': {
+            'type': 'private',
+            'to': "foo_receiver@zulip.com",
+        },
+        'stream': {
+            'type': 'stream',
+            'to': 'foo',
+            'subject': 'foo',
+        },
+    }
+
+    def check_expected_responses(self, content_expectations,
+                                 message_templates=list(message_template.values()),
+                                 default_method='send_reply',
+                                 default_response_template={},
+                                 default_result={'result': 'success', 'id': 5}):
+        # type: (Dict[str, Any], List[Dict], str, Dict, Dict) -> None
+        # To use defaults, Any would be a str type.
+        # To use non-defaults, Any would be a tuple.
+        for mt in message_templates:
+            for m, r in content_expectations.items():
+                resp = r if isinstance(r, list) else [r]
+                # Expand message content to have current message template
+                message = dict(mt, content = m)
+                # Check if entries are tuples they are correct form
+                for i in resp:
+                    if isinstance(i, tuple):
+                        li = len(i)
+                        if li != 3 and li != 2:
+                            logging.critical("tuple must be length 2 or 3 (or just content)")
+                        elif li == 3 and i[1] != 'send_message':
+                            logging.exception("send_message requires message/'send_message'/response_template")
+                        elif li == 2 and i[1] != 'send_reply':
+                            logging.exception("send_reply requires message/'send_reply' only")
+                    else:
+                        if default_method == 'send_message' and default_response_template == {}:
+                            raise Exception("send_message default method requires default_response_template")
+                # Expand each response to use default template, unless specified in entry
+                # Result is a list of (response, method, result) pairs
+                # FIXME Small change, but cannot specialise result to be not a success in table
+                responses = [((dict(i[2], content = i[0]), i[1], default_result) if len(i) == 3 else (i[0], i[1], default_result))
+                             if isinstance(i, tuple)
+                             else (dict(default_response_template, content = i), default_method,
+                                   default_result)
+                             for i in resp]
```

Comment:
I'm not sure if there is a better comprehension syntax for this.

The updated code is clearer than the original, I believe.

The only option I can think of at the moment is doing multiple comprehensions and putting the separate 'if' statements as filters, but that wouldn't preserve the order of the original list.